ir/inode.h

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef IR_INODE_H_
+#define IR_INODE_H_
+
+#include "lib/castable.h"
+#include "lib/cstring.h"
+#include "lib/exceptions.h"
+#include "lib/rtti.h"
+#include "lib/source_file.h"
+
+namespace P4 {
+class JSONGenerator;
+class JSONLoader;
+}  // namespace P4
+
+namespace P4::IR {
+
+class Node;
+
+/// SFINAE helper to check if given class has a `static_type_name`
+/// method. Definite node classes have them while interfaces do not
+template <class, class = void>
+struct has_static_type_name : std::false_type {};
+
+template <class T>
+struct has_static_type_name<T, std::void_t<decltype(T::static_type_name())>> : std::true_type {};
+
+template <class T>
+inline constexpr bool has_static_type_name_v = has_static_type_name<T>::value;
+
+// node interface
+class INode : public Util::IHasSourceInfo, public IHasDbPrint, public ICastable {
+ public:
+    virtual ~INode() {}
+    virtual const Node *getNode() const = 0;
+    virtual Node *getNode() = 0;
+    virtual void toJSON(JSONGenerator &) const = 0;
+    virtual cstring node_type_name() const = 0;
+    virtual void validate() const {}
+
+    // default checkedTo implementation for nodes: just fallback to generic ICastable method
+    template <typename T>
+    std::enable_if_t<!has_static_type_name_v<T>, const T *> checkedTo() const {
+        return ICastable::checkedTo<T>();
+    }
+
+    // alternative checkedTo implementation that produces slightly better error message
+    // due to node_type_name() / static_type_name() being available
+    template <typename T>
+    std::enable_if_t<has_static_type_name_v<T>, const T *> checkedTo() const {
+        const auto *result = to<T>();
+        BUG_CHECK(result, "Cast failed: %1% with type %2% is not a %3%.", this, node_type_name(),
+                  T::static_type_name());
+        return result;
+    }
+
+    DECLARE_TYPEINFO(INode);
```

Comment:
@asl I have a question on the RTTI macros. I am trying to factor out some classes to avoid pulling in the entire IR code. Unfortunately, for IDeclaration and INode macros like  `DECLARE_TYPEINFO_WITH_TYPEID(INode, NodeKind::INode);` require pulling in `ir/ir-tree-macros.h` which is generated.

Is there a way to avoid this while still being correct?

Reply:
Well, the typeid's should be unique. It is part of `gen-tree-macro.h`. The location is a bit weird, but it was emitted there exactly in order not to pull the whole IR header. It could be even split into separate .h file if `gen-tree-macro.h` is undesired.

Reply:
Yeah, gen-tree-macro depends on the generator which is a little unpleasant. Although the code there is less complex than what is being put into ir-generated.h.
You effectively run into https://github.com/p4lang/p4c/issues/5013. 

Maybe there is a way to depend on gen-tree or just the rtti enum.

Reply:
Well, one way to do this is to split enum into "static" and "generated" parts. Static would be for all classes not autogenerated (`INode`, `IDeclaration`, `Vector`), etc. and everything else will be outside the scope.